api/http/management.go

```diff
@@ -408,67 +409,84 @@ func (h ManagementController) websocketWriter(
 	lastKeystrokeAt := time.Now().UTC().UnixNano()
 Loop:
 	for {
+		var forwardedMsg []byte
+
 		select {
 		case msg := <-deviceChan:
 			mr := &ws.ProtoMsg{}
 			err = msgpack.Unmarshal(msg.Data, mr)
 			if err != nil {
 				return err
 			}
+
+			forwardedMsg = msg.Data
+
 			if mr.Header.Proto == ws.ProtoTypeShell {
 				switch mr.Header.MsgType {
 				case shell.MessageTypeShellCommand:
-					timeNowUTC := time.Now().UTC().UnixNano()
-					keystrokeDelay := timeNowUTC - lastKeystrokeAt
-					if keystrokeDelay >= keyStrokeDelayRecordingThresholdNs {
-						if keystrokeDelay > keyStrokeMaxDelayRecording {
-							keystrokeDelay = keyStrokeMaxDelayRecording
+
+					if recordedBytes >= app.MessageSizeLimit || controlBytes >= app.MessageSizeLimit {
+						sendLimitErrDevice(ctx, session, h.nats)
+						userErrMsg, err := prepLimitErrUser(ctx, session)
+						if err != nil {
+							l.Errorf("session limit: " +
+								"failed to notify user")
 						}
-						controlMsg := app.Control{
-							Type:   app.DelayMessage,
-							Offset: recordedBytes,
-							DelayMs: uint16(float64(keystrokeDelay) *
-								0.000001),
-							TerminalHeight: 0,
-							TerminalWidth:  0,
+
+						//override original message with shell error
+						forwardedMsg = userErrMsg
+					} else {
+						// session recording
+						b, e := recorderBuffered.Write(mr.Body)
+						if e != nil {
+							l.Errorf("session logging: "+
+								"recorderBuffered.Write"+
+								"(len=%d)=%d,%+v",
+								len(mr.Body), b, e)
 						}
-						n, _ := controlRecorderBuffered.Write(
-							controlMsg.MarshalBinary())
-						l.Debugf("saving control delay message: %+v/%d",
-							controlMsg, n)
-						controlBytes += n
-						if controlBytes >= app.MessageSizeLimit {
-							l.Infof("closing session session_id=%s"+
-								"on control data limit reached.",
-								session.ID)
```

Comment:
so, the fact that we are closing the shell session is not logged now?

Reply:
mistake, will fix

Reply:
actually, the logging was c+v twice, so it's still there.